### PR TITLE
Reworked main pipeline for ci tests

### PR DESCRIPTION
**Pull request by starkmsu**
CI tests improvement

**Description**: Reworked the pipeline for CI tests.
List of improvements:
1. Extracted tfx init steps from PreDelete, Test Upload and PostDelete stages into a separate step to avoid repeating runs of the same commands
2. Removed sources checkout from PreDelete, Test Upload and PostDelete since it can be done only once at the start of Build stage
3. Processing of different tasks is made parallel - 1 task on only 1 build agent when there is no need to publish hotfix artifact
4. When there is a need to publish hotfix artifact all tasks are processed on a single build agent
5. Steps to publish nuget package are made optional and dependent on PushNugetPackageflag in the main pipeline
6. Steps for building hotfix artifact are made optional and dependent on PublishHotfixArtifact in the main pipeline 

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** (Y/N) N

**Checklist**:
- [ ] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected

### CODE REVIEW

**Review by z00sts — APPROVED** [review]

File: ci/ci-test-tasks/main.yml [hunk 1]
```diff
@@ -6,78 +6,103 @@ parameters:
 - name: BuildSourceVersion
   displayName: The latest version control change of the triggering repo that is included in this build.
   type: string
+- name: PushNugetPackage
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/main.yml:9]: We should remove the parameters PushNugetPackage and PublishHotfixArtifact at all since they always have the same value. The original sources shows: 
 parameters:
            task: ${{ task }}
            push: false
            publishArtifact: true
            patchVersion: true
            version: '0.999.0'
            BuildSourceVersion: ${{ parameters.BuildSourceVersion }}
         
So I suggest just get rid of the push, publishArtifact and patchVersion parameters and any corresponding code that depends on it. 
It seems it was copied from another pipeline definition where those parameters are actually used.

**Review by PavloAndriiesh — CHANGES_REQUESTED** [review]

File: ci/ci-test-tasks/build-task.yml [hunk 2]
```diff
@@ -27,26 +21,25 @@ steps:
 - powershell: ./ci/start-collect-diagnostics.ps1
   displayName: Start collect diagnostics
 
-# Use node 8, npm 5
+# Use node 10, npm 6
```
[the remark this answers is not in the document]
Comment (starkmsu) [reply ci/ci-test-tasks/build-task.yml:24]: Here it's said to use Node 10 - https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md#prerequisites-node-and-npm

[the remark this answers is not in the document]
Comment (starkmsu) [reply ci/ci-test-tasks/build-task.yml]: Restored Node 8 usage

**Review by LeftTwixWand — APPROVED** [review]

File: ci/ci-test-tasks/main.yml (hunk 1 above)
Comment (starkmsu) [reply ci/ci-test-tasks/main.yml:9]: These steps are not always needed that's why I introduced flags for them. In most cases nuget pushing and hotfix artifact uploading are not required.

Comment (PavloAndriiesh) [reply ci/ci-test-tasks/main.yml:9]: I think I didn't comment on some parts clearly. What I mean is we do not need them at all since they are not changing. It is constant:
push: false
publishArtifact: true	
patchVersion: true

According to this we can get rid of them at all.

**Review by PavloAndriiesh — CHANGES_REQUESTED** [review]

File: ci/ci-test-tasks/build-init.yml [hunk 3]
```diff
@@ -1,23 +1,31 @@
 parameters:
-  - name: task
-    type: string
-  - name: BuildSourceVersion
-    type: string
+- name: BuildSourceVersion
+  type: string
 
 steps:
-- checkout: none
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/build-init.yml:8]: Why did you delete the "checkout: none" step? It will checkout the current master branch and then override it with the checkout of BuildSourceVersion. Can take longer time

File: ci/ci-test-tasks/main.yml [hunk 4]
```diff
@@ -6,78 +6,150 @@ parameters:
...
-  jobs:
-    - ${{ each task in split(parameters.tasks, ',') }}:
-      - job: ${{ task }}_test_task
-        displayName: Test (${{ task }})
+          # Stop collect diagnostics
+          - powershell: ./ci/stop-collect-diagnostics.ps1
+            displayName: Stop collect diagnostics
+            condition: always()
+
+          - script: |
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/main.yml:49]: Here for each task we install tfx-cli and login? I am afraid it will take way to long on PR's with a number of tasks. Some time we have PRs with tens of tasks changed.

File: ci/ci-test-tasks/main.yml [hunk 5]
```diff
@@ -6,78 +6,150 @@ parameters:
...
+
+          - ${{ if parameters.PushNugetPackage }}:
+            - template: ./push-nuget-init.yml
+
+            - script: |
+                cd $(Build.SourcesDirectory)\_package\nuget-packages\$(task)
+                push.cmd
+              displayName: Push Nuget package for $(task)
+
+    - ${{ if parameters.PublishHotfixArtifact }}:
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/main.yml:86]: This part is redundant. We should never use PublishHotfixArtifact in this pipeline. It's a part of another pipeline where we make hotfixes. PublishHotfixArtifact is always false.

File: ci/ci-test-tasks/push-nuget-init.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# requires some rework - relies on batch processing and steps are not self-evindent
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/push-nuget-init.yml:1]: All the nuget functionality is also redundant since this pipeline should never publish to nuget.

File: ci/ci-test-tasks/upload-hotfix.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+parameters:
```
Comment (PavloAndriiesh) [inline ci/ci-test-tasks/upload-hotfix.yml:1]: Same with the hotfix - it is never used here. The hotfix functionality is used in another pipeline. We can get rid of it

**Review by PavloAndriiesh — CHANGES_REQUESTED** [review]

File: ci/ci-test-tasks/main.yml (hunk 1 above)
Comment (starkmsu) [reply ci/ci-test-tasks/main.yml:9]: Most of this pipeline runs are to test new PRs, for such case we don't need to push nuget packages and publish artifacts. We only need to patch version of the task and test it with ci tests. So configuration of flags will be like this:
push: false, publishArtifact: false, patchVersion: true
When we need a hotfix we need the following configuration of flags:
push: true, publishArtifact: true, patchVersion: false
For a rare case when we don't need a hotfix but only a nuget package for some reason flags will be the following:
push: true, publishArtifact: false, patchVersion: false
So as you can see we do need PushNugetPackage and PublishHotfix flags since there multiple permutations of those.
patchVersion might be resolve as not(push) though

File: ci/ci-test-tasks/main.yml (hunk 4 above)
Comment (starkmsu) [reply ci/ci-test-tasks/main.yml:49]: Yes, but each task is being executed on a different build agent since it's a separate job, so each build machine has to have its own version of tfx installed to proceed.

File: ci/ci-test-tasks/main.yml (hunk 5 above)
Comment (starkmsu) [reply ci/ci-test-tasks/main.yml:86]: Aha, that simplifies things a lot! Let me remove this part.

File: ci/ci-test-tasks/push-nuget-init.yml (hunk 6 above)
Comment (starkmsu) [reply ci/ci-test-tasks/push-nuget-init.yml:1]: Got it

File: ci/ci-test-tasks/upload-hotfix.yml (hunk 7 above)
Comment (starkmsu) [reply ci/ci-test-tasks/upload-hotfix.yml:1]: Will do

File: ci/ci-test-tasks/main.yml (hunk 1 above)
Comment (starkmsu) [reply ci/ci-test-tasks/main.yml:9]: After reading other comments I've realized that we don't need hotfix or nuget push functionality here. So I will remove these flags then.

**Review by kirill-ivlev — APPROVED** [review]

**Review by ozanaydinMSFT — CHANGES_REQUESTED** [review]
image

I've talked with @starkmsu . Parallel run with tasks of same id will cause issue. 
Since delete operation is deleting all versions, order should be "delete => upload => test => postdelete" for all.

(e.g. if AzureFileCopyV1's "delete" works at the end of "AzureFileCopyV2" 's "upload" stage;  "AzureFileCopyV2" 's "test" stage will work with organization's task version (not PR's version))

**Review by ozanaydinMSFT — APPROVED** [review]
After revision for same id tasks we tested again. Performance tests are better for many situation. Approximately similar for same id parallel tasks runs. 

In future, we can consider to use "common" tools (node, tfx, axios etc.) to reduce time for parallel jobs. But for now this version is better to use.

**Review by PavloAndriiesh — APPROVED** [review]
LGTM but still can have some hidden issues. @starkmsu Please keep an eye on a ci pipelines during the next weeks after merging.
